### PR TITLE
Implemented file reading support for Go code generator

### PR DESCRIPTION
**Pull request by T-Vova**
# Description
This system enables any function call to be given its own custom helper method positioned in between the imports and the main function. Since reading a file in Go requires error handling code, having it be encapsulated in a helper method helps reduce repetition and improve overall understandability especially if many files are read.

The following HCL example from the AWS provider, along with 12 others, are now able to successfully convert into go.
Before:
```hcl
resource "aws_pinpoint_apns_voip_channel" "apns_voip" {
  application_id = aws_pinpoint_app.app.application_id

  certificate = file("./certificate.pem")
  private_key = file("./private_key.key")
}

resource "aws_pinpoint_app" "app" {}
```

After:
```go
package main

import (
	"io/ioutil"

	"github.com/pulumi/pulumi-aws/sdk/v4/go/aws/pinpoint"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

func readFileOrPanic(path string) pulumi.StringPtrInput {
	if fileData, err := ioutil.ReadFile(path); err == nil {
		return Pulumi.String(string(fileData[:]))
	} else {
		panic(err.Error())
	}
}

func main() {
	pulumi.Run(func(ctx *pulumi.Context) error {
		app, err := pinpoint.NewApp(ctx, "app", nil)
		if err != nil {
			return err
		}
		_, err = pinpoint.NewApnsVoipChannel(ctx, "apnsVoip", &pinpoint.ApnsVoipChannelArgs{
			ApplicationId: app.ApplicationId,
			Certificate:   readFileOrPanic("./certificate.pem"),
			PrivateKey:    readFileOrPanic("./private_key.key"),
		})
		if err != nil {
			return err
		}
		return nil
	})
}
```

### CODE REVIEW

**Review by lukehoban — COMMENTED** [review]
This is a nice approach to supporting these functions in Go!  

It's a slight shame that we have to emit helpers that panic (which is not really idiomatic or recommended practice), but I can't see any simple way for us to avoid that, and I suspect the net increase in complexity needed to support pushing this into normal error handling flow wouldn't be worth it here. 

One suggestion on a tweak to the helper function implementation. 

It would be great to have test coverage of these new codepaths in this library.

File: pkg/codegen/go/gen_program_utils.go [hunk 1]
```diff
@@ -54,3 +54,26 @@ func (p *promptToInputArrayHelper) getPromptItemType() string {
...
+
+	switch functionName {
+	case "readFile":
+		methodBody =
+			`func readFileOrPanic(path string) pulumi.StringPtrInput {
+				if fileData, err := ioutil.ReadFile(path); err == nil {
+					return pulumi.String(string(fileData[:]))
+				} else {
+					panic(err.Error())
+				}
```
[the remark this answers is not in the document]
Comment (T-Vova) [reply pkg/codegen/go/gen_program_utils.go:69]: Done.

File: pkg/codegen/go/gen_program_utils.go [hunk 2]
```diff
@@ -54,3 +54,26 @@ func (p *promptToInputArrayHelper) getPromptItemType() string {
...
+			}`
```
[the remark this answers is not in the document]
Comment (T-Vova) [reply pkg/codegen/go/gen_program_utils.go]: Done.

**Comment by pgavlin** [comment]
> It's a slight shame that we have to emit helpers that panic (which is not really idiomatic or recommended practice), but I can't see any simple way for us to avoid that, and I suspect the net increase in complexity needed to support pushing this into normal error handling flow wouldn't be worth it here.

We could instead emit helpers that return a `pulumi.StringOutput` that is either resolved (in the case that there is no error) or rejected (in the case that there is an error). This might look like:

```go
func readFile(ctx *pulumi.Context, path string) pulumi.StringInput {
	output, resolve, reject := ctx.NewOutput()
	if fileData, err := ioutil.ReadFile(path); err == nil {
		resolve(fileData)
	} else {
		reject(err)
	}
	return output.ApplyT(func(v interface{}) string { return v.(string) }).(pulumi.StringOutput)
}
```

**Comment by lukehoban** [comment]
> We could instead emit helpers that return a pulumi.StringOutput

On balance, I think I prefer the original proposal @T-Vova has here (or more precisely, the slight tweak on that after some updates in this PR).  It's simpler, and honestly the error handling is more "obvious" so easier to adapt to what the user might really want to do if they do want to do something different.

**Comment by pgavlin** [comment]
> On balance, I think I prefer the original proposal @T-Vova has here (or more precisely, the slight tweak on that after some updates in this PR). It's simpler, and honestly the error handling is more "obvious" so easier to adapt to what the user might really want to do if they do want to do something different.

I can see wanting to avoid some of the gritty bits of manually-resolved outputs. We can stick with the original approach, though I really don't like the use of `panic`, as it isn't idiomatic for Go or Pulumi.

**Comment by T-Vova** [comment]
One possible compromise could be to have the helper method contain a comment right above it linking to a tutorial on using pulumi inputs/outputs, saying something like:
```ts
// Visit pulumi.com/docs/intro/concepts/inputs-outputs/ to learn more about 
// manually-resolving outputs and how to avoid using panic() when handling errors
```
 This would keep the example simple to understand, but would also clearly indicate where the user could go to in order to learn about the correct approach.

**Comment by stack72** [comment]
I am going to go ahead and merge this - it has been decided to be the best way forward :) The failures are unrelated here

**Review by stack72 — APPROVED** [review]

**Pull request merged**